app/Traits/HasPointsAttribute.php

```diff
@@ -29,18 +29,24 @@ public function validatePoints(array $points)
                 $valid = $size === 2;
                 break;
             case Shape::circleId():
-                $valid = $size === 3;
+                $valid = $size === 3 && intval(($points)[2]) !== 0;
                 break;
             case Shape::ellipseId():
             case Shape::rectangleId():
-                $valid = $size === 8;
+                $valid = $size === 8 && $this->countDistinctCoordinates($points) === 4;
+                break;
+            case Shape::polygonId():
+                $valid = $size % 2 === 0 && count($points) >= 6 && $this->countDistinctCoordinates($points) >= 3;
```

Comment:
This must be:
```suggestion
                $valid = $size % 2 === 0 && count($points) >= 8 && $this->countDistinctCoordinates($points) >= 3;
```
I think. It can't have only 3 points because the first must equal the last and we no longer allow a line here, right?